src/Interpreters/GraceHashJoin.cpp

```diff
@@ -756,6 +760,20 @@ size_t GraceHashJoin::getNumBuckets() const
     return buckets.size();
 }
 
+bool GraceHashJoin::rightTableCanBeReranged() const
+{
+    if (hash_join && buckets.size() <= 1)
```

Comment:
based on two methods below shouldn't we take `hash_join_mutex` and `rehash_mutex`?

Reply:
take `rehash_mutex` by use `getNumBuckets`, and I think we should not take `hash_join_mutex` for call `hash_join->rightTableCanBeRereanged()`, as this method access the variants in `hash_join` with no data race.